sdk/program/src/loader_v4.rs

```diff
@@ -41,6 +46,130 @@ impl LoaderV4State {
     }
 }
 
+pub fn is_write_instruction(instruction_data: &[u8]) -> bool {
+    !instruction_data.is_empty() && 0 == instruction_data[0]
+}
+
+pub fn is_truncate_instruction(instruction_data: &[u8]) -> bool {
+    !instruction_data.is_empty() && 1 == instruction_data[0]
+}
+
+pub fn is_deploy_instruction(instruction_data: &[u8]) -> bool {
+    !instruction_data.is_empty() && 2 == instruction_data[0]
+}
+
+pub fn is_retract_instruction(instruction_data: &[u8]) -> bool {
+    !instruction_data.is_empty() && 3 == instruction_data[0]
+}
+
+pub fn is_transfer_authority_instruction(instruction_data: &[u8]) -> bool {
+    !instruction_data.is_empty() && 4 == instruction_data[0]
+}
+
+/// Returns the instructions required to initialize a program/buffer account.
+pub fn create_buffer(
+    payer_address: &Pubkey,
+    buffer_address: &Pubkey,
+    lamports: u64,
+) -> Instruction {
+    system_instruction::create_account(payer_address, buffer_address, lamports, 0, &id())
+}
+
+/// Returns the instructions required to set the length of the program account.
+pub fn truncate(
+    program_address: &Pubkey,
+    authority: &Pubkey,
+    new_size: u32,
+    recipient_address: &Pubkey,
+) -> Instruction {
+    Instruction::new_with_bincode(
+        id(),
+        &LoaderV4Instruction::Truncate { new_size },
+        vec![
+            AccountMeta::new(*program_address, true),
```

Comment:
Here, the program does not need to sign. Only in the `Truncate` inside `create_buffer()` the program needs to be a signer to set the `authority` initially.